[Split Build] Load dependencies from libtorch in __init__.py

PaliC opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #126813
* #126699
* __->__ #126826

This PR makes it such that we search for a libtorch wheel when initializing pytorch in order to find the necessary shared libraries.

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update on "[Split Build] Load dependencies from libtorch in __init__.py"


This PR makes it such that we search for a libtorch wheel when initializing pytorch in order to find the necessary shared libraries.


[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

huydhn commented on tools/setup_helpers/env.py:
```diff
@@ -21,6 +21,8 @@
 
 BUILD_DIR = "build"
 
+LIBTORCH_PKG_NAME = "libtorchsplit"
```
I wish we have a better name for this, i.e. libtorch-lite?

ZainRizvi replied:
@PaliC can we give it a name more descriptive of what it's supposed to do? 

E.g. Is this a wrapper around libtorch? Is it the core libtorch library?

PaliC replied:
Yeah this is a placeholder name. I was planning on having a better name in a separate PR after we decide on one. Sadly we gotta look for an open pypi package for it.

huydhn commented on torch/__init__.py:
```diff
@@ -165,11 +167,57 @@ def _preload_cuda_deps(lib_folder, lib_name):
+                pass
+        return None
+
+    def load_shared_libraries(library_path):
```
A curious q:  The `ctypes.CDLL` call below online 179 doesn't support Windows, but this function does.  What would be the reason for this discrepancy?

ZainRizvi commented on torch/__init__.py:
```diff
@@ -165,11 +167,57 @@ def _preload_cuda_deps(lib_folder, lib_name):
+            lib_ext = '.so'
+
+        # Find all shared library files with the appropriate extension
+        so_files = [f for f in os.listdir(lib_dir) if f.endswith(lib_ext)]
```
nit: since the code doesn't. only support .so files

```suggestion
        library_files = [f for f in os.listdir(lib_dir) if f.endswith(lib_ext)]
```

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update on "[Split Build] Load dependencies from libtorch in __init__.py"


This PR makes it such that we search for a libtorch wheel when initializing pytorch in order to find the necessary shared libraries.


[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC committed:
Update

[ghstack-poisoned]

PaliC commented:
@pytorchbot merge

PaliC commented:
@pytorchbot merge

PaliC committed:
Update

[ghstack-poisoned]

PaliC commented:
@pytorchbot merge

PaliC commented:
@pytorchbot merge

PaliC committed:
Update

[ghstack-poisoned]

PaliC commented:
@pytorchbot merge

PaliC commented:
@pytorchbot merge

PaliC commented:
@pytorchbot merge

PaliC commented:
@pytorchbot merge

PaliC commented:
@pytorchbot merge

PaliC committed:
Update

[ghstack-poisoned]

PaliC commented:
@pytorchbot merge
